### PR TITLE
Check if angles in pose are within range as set in 'settings.py'.

### PR DESCRIPTION
**Pull request by markkulube**

### CODE REVIEW

**Comment by markkulube** [comment]
@mithi I am not sure if i can address the cognitive complexity as well as the failed build....i may need help with both of these...

also, the lower bound of the angles is zero right?

File: hexapod/models.py [hunk 1]
```diff
@@ -240,6 +243,12 @@ def _update_local_frame(self, frame):
 # Helper functions
 # ..........................................
 
+def poses_within_range(pose):
+    # Check if any angle value in poses is out of range
+    if(pose["coxia"] < 0 or pose["coxia"] > ALPHA_MAX_ANGLE or pose["femur"] < 0 or pose["femur"] > BETA_MAX_ANGLE or pose["tibia"] < 0 or pose["tibia"] > GAMMA_MAX_ANGLE):
```
Comment (mithi) [inline hexapod/models.py]: The range should be: ` -MAX_ANGLE <= angle <= MAX_ANGLE`

**Comment by mithi** [comment]
Hi @markkulube , 

Unfortunately, I won't be able to merge your PR as there are quite a few misunderstanding we have! I hope you don't take it personally! I'm sorry that I won't be able to merge your pull request. I hope that this is not a bad experience for you. I understand if you think this is too much work. Feel free abandon this PR if so. 


Below are some of my thoughts

1. `-MAX_ANGLE < angle < MAX_ANGLE` is the range. You would have noticed that this is the range of the sliders when you play around the app. 

2. You can run an automatic code formatter https://github.com/psf/black to make sure your code style is up to par
All the tests should passed, take a look at travis to find out where exactly it fails. 

3. `check_pose_within_range()` should be called before any other operation in that function ie it should be called right after `def update`

When structuring your code, always try to put the one reading your code in mind. Ask yourself, what would make my code so much easier to understand? What is this code trying to achieve? 

If you haven't yet, pep20 and pep8 are really good guidelines (guidelines, not rules!)
- https://www.python.org/dev/peps/pep-0020/
- You can use pep 8 https://www.python.org/dev/peps/pep-0008/ and

The blocks of code below is something you can explore, but it's probably not the best, you can read a few blog articles and maybe let me know what you think could be the best way to approach the problem at hand. The code below is something
you can explore, but I haven't given much thought to it. This is why I didn't implement this code myself.

```python
def check_pose_within_range(poses):
    for leg_name, pose in pose.items():
        alpha, beta, gamma = pose['coxia], pose['femur'], pose['tibia]
        alpha_in_range = -MAX_ALPHA < alpha < MAX_ALPHA
        if not alpha_in_range: 
           msg = f"{leg_name} leg's alpha = {alpha} is not within [-{MAX_ALPHA}, {MAX_ALPHA}]"
            raise Exception(msg)
        ....        
```

Again, I'm sorry if i mislead you. I'm also learning from this experience on how to communicate better.

File: hexapod/models.py [hunk 2]
```diff
@@ -240,6 +243,15 @@ def _update_local_frame(self, frame):
 # Helper functions
 # ..........................................
 
+def poses_within_range(pose):
+    # Check if any angle value in poses is out of range
+    coxia_in_range = pose["coxia"] < 0 or pose["coxia"] > ALPHA_MAX_ANGLE
```
Comment (mithi) [inline hexapod/models.py:248]: I think this is `coxia_out_of_range` not `coxia_in_range` don't you think?

**Comment by markkulube** [comment]
Hi @mithi,

Thanks again for the very detailed feedback.

Please don't be hard on yourself. I think you are being a great maintainer/owner.

To be honest, this is pretty much everyday at work i.e. constant back and forth work about code quality, algorithms, and implementation etc.  For instance, yesterday I raised a PR at work, my boss had a couple comments, I fixed them, and my feature was merged into DEV. Not just work but I also experience this in other open source projects as well. So please, as the project maintainer/owner I expect you to give code reviews so that PRs meet the standards of your project.

Yeah, my initial PR was kinda sloppy - sorry about that. It should be much much cleaner and better refactored now in the latest commit.

Let me respond to some of your points:
1. Please allow project owners and maintainers to edit your pull request. Here's the work in progress contributing guide I'm developing.
MARK: Yes I always check the box "Allow edits by maintainers", as it is now - are you not able to edit my PR?

2. -MAX_ANGLE < angle < MAX_ANGLE is the range. You would have noticed that this is the range of the sliders when you play around the app.
***MARK: True, but this should be obvious in the back end implementation of the project. The static/global variables should be explicitly defined. This is the second time you are leaving things to inference. This is not necessarily a bad practice, but as your project grows, some developers may only want to contribute to the backend without ever seeing the UI components. I have experienced this another  OSS project where contributors can work on either part independently. In setting.py consider defining something like:
ALPHA_MAX_ANGLE={"ALPHA_LOWER": -90, "ALPHA_UPPER": 90}
BETA_MAX_ANGLE={"BETA_LOWER": -180, "BETA_UPPER": 180}
GAMMA_MAX_ANGLE={"GAMMA_LOWER": -180, "GAMMA_UPPER": 180}
This is how I would do it since later it gives me the power define the lower and upper bounds independently of each other. Right now the lower bound is too dependent of the upper bound.

3. You can run an automatic code formatter https://github.com/psf/black to make sure your code style is up to par
All the tests should passed, take a look at travis to find out where exactly it fails.
***MARK: Okay, will at the travis logs then.

4. check_pose_within_range() should be called before any other operation in that function ie it should be called right after def update
***MARK: Yeah, I did just that, but why is this case?

5. When structuring your code, always try to put the one reading your code in mind. Ask yourself, what would make my code so much easier to understand? What is this code trying to achieve?
***MARK: Yes, definitely, I realize I was a kinda sloppy with the initial PR. I think readability should always be a top priority when writing code, cause we want the developer to make sense of it without much hassle!

Again, I'm sorry if i mislead you. I'm also learning from this experience on how to communicate better.
***MARK: No need to apologize, again this really is work flow I am used to in development. In terms of giving code reviews, you are be nice :). Again, I expect you as the project lead to be blunt for the sake of keeping up your project standards.

 I understand if you think this is too much work. Feel free abandon this PR if so.
***MARK: Nah, I love doing this especially seeing projects, even as small as this to the end!

Let me know what you think of the most recent commit!

Thanks again, @mithi

File: hexapod/models.py [hunk 3]
```diff
@@ -117,43 +117,43 @@ def __init__(self, dimensions):
...
     def update(self, poses, assume_ground_targets=True):
-        self.body_rotation_frame = None
-        might_twist = find_if_might_twist(self, poses)
-        old_contacts = deepcopy(self.ground_contacts)
-
-        # Update leg poses
-        for pose in poses.values():
-            i = pose["id"]
-            if(poses_within_range(pose)):
+        if(poses_within_range(poses)):
```
Comment (mithi) [inline hexapod/models.py:120]: I prefer exiting early. as the zen of python says, flat is better than nested. 

IE

```
if poses_within_range(poses): 
    return

// do what you want to do here

**Comment by mithi** [comment]
I'm sorry I don't have time to oversee this commit because of work. So I may not be able to reply any time soon.

**Comment by mithi** [comment]
@markkulube Thank you for allowing me to edit this PR. I'll attend to this at the soonest possible time. Thanks again!

File: hexapod/models.py [hunk 4]
```diff
@@ -241,6 +242,37 @@ def _update_local_frame(self, frame):
 # ..........................................
 
 
+def poses_within_range(poses):
+    """ Return true if all angles of all legs are within range of  [-MAX_ANGLE, MAX_ANGLE]"""
+
+    def _angle_range_exception(angle_in_range, angle, angle_name, leg_name, max_angle, all_in_range):
```
Comment (mithi) [inline hexapod/models.py:248]: Some errors said by travis

```
hexapod/models.py:248:0: C0301: Line too long (101/100) (line-too-long)

hexapod/models.py:252:0: C0301: Line too long (102/100) (line-too-long)

hexapod/models.py:270:0: C0301: Line too long (103/100) (line-too-long)

hexapod/models.py:272:0: C0301: Line too long (103/100) (line-too-long)

```

**Comment by mithi** [comment]
@markkulube Thank you for allowing me to edit this PR. I'll make the necessary changes and then I'll finally squash and merge this PR. I'll attend to this at the soonest possible time. Thanks again!

**Comment by markkulube** [comment]
@mithi i think i resolved everything now in terms of the checks.

so what's really left is double checking the implementation of def poses_within_range(poses), but beyond that nothing should be in the way to merge and close this PR.

**Comment by markkulube** [comment]
@mithi that was sublime refactor.

**Pull request merged**